### PR TITLE
flavor_settings field within each Incidents flavor

### PR DESCRIPTION
**Pull request by mpagot**
Add support for a new Incidents configuration field that allow to
specify flavor specific settings.

Keep separated settings for each flavor loop

Ticket: TEAM-8180

### CODE REVIEW

File: openqabot/types/incidents.py
Comment (mpagot) [inline openqabot/types/incidents.py:103]: We can potentially further optimize the code in regard of this copy and the one in 
https://github.com/openSUSE/qem-bot/pull/145/files#diff-81a88cee51bd2524a7e938a6b6d53d79f526ffb436071ead2d3d83163b38b1ecR255

File: openqabot/types/incidents.py
Comment (ricardobranco777) [inline openqabot/types/incidents.py:141]:
```suggestion
                    # Settings for the POST command
                    full_post["openqa"].update(flavor_settings)
```

The assignment after the `.update()` is redundant.

File: openqabot/types/incidents.py
Comment (ricardobranco777) [inline openqabot/types/incidents.py:143]:
```suggestion
```

Also redundant.

**Review by ricardobranco777 — CHANGES_REQUESTED** [review]

**Comment by Martchus** [comment]
The first commit message should likely start with "Add …". If there's a related ticket then it would also be good to reference it in the commit message bodies.

**Comment by mpagot** [comment]
> The first commit message should likely start with "Add …". If there's a related ticket then it would also be good to reference it in the commit message bodies.

PR has intentionally two commits. There's a bug in the first one that I fixed in the second one, but I like, at least reviewer, to be aware about it.

Bug is potentially also related to a possible further optimization in the code in regard of

```
settings = full_post["openqa"].copy()
```

Please have a look and let me know what do you think.

I think that this double commit is more interesting from the point of view of the review. So maybe I can squash them at the moment of merge.  What do you think?

**Comment by okurz** [comment]
You don't need to squash commits if they provide a reasonable story each by themselves. We commonly follow http://chris.beams.io/posts/git-commit/ so please take a look into that blog article if you don't know it yet. At the very least rephrase the git commit message subjects to start in imperative mood starting with capital letter as Martchus stated.

File: openqabot/types/incidents.py
Comment (asmorodskyi) [inline openqabot/types/incidents.py:107]: how this related to introduction of `flavor_settings` ?

File: openqabot/types/incidents.py
Comment (asmorodskyi) [inline openqabot/types/incidents.py:111]: how this is related to introduction of flavor_settings ?

File: openqabot/types/incidents.py
Comment (asmorodskyi) [inline openqabot/types/incidents.py:109]: disagree , flavor specific settings should NOT allow to overwrite VERSION defined on product level , this will open gate for so many tricky bugs and data inconsistencies that we do NOT want to deal with

File: openqabot/types/incidents.py
Comment (asmorodskyi) [inline openqabot/types/incidents.py:103]: IMHO all changes what you have done above and below may be replaced by just TWO lines added here ( and we NOT removing this one too ofc. ) 

```
 if "flavor_settings" in data:
                 full_post["openqa"].update(data["flavor_settings"])
```

to be on safe side we may add verification something like : 
```
if "VERSION" in data["flavor_settings"] or "DISTRI" in data["flavor_settings"]:
                    raise ValueError("Attempt to redifine VERSION/DISTRI on flavor level")
```

Comment (mpagot) [reply openqabot/types/incidents.py:103]: I'm ok if it pass the UT I created. I'm giving it a try

**Comment by mpagot** [comment]
> The first commit message should likely start with "Add …".

Done

File: openqabot/types/incidents.py
Comment (mpagot) [reply openqabot/types/incidents.py:141]: > The assignment after the .update() is redundant.

Was there even before this PR. I'll address it in a next PR

File: openqabot/types/incidents.py
Comment (mpagot) [reply openqabot/types/incidents.py:143]: > Also redundant.

Thanks, good catch but was there even before this PR. I'll address it in a next PR

File: openqabot/types/incidents.py
Comment (mpagot) [reply openqabot/types/incidents.py:107]: This code has just be moved not added. 

As `flavor_settings` is updating `settings` I reorganized the code to have all stuff with different purpose as separated and sequentially ordered as possible. I can recognize at least 3 main category.

* check something and skip to schedule a POST, me moving https://github.com/openSUSE/qem-bot/pull/145/files#diff-81a88cee51bd2524a7e938a6b6d53d79f526ffb436071ead2d3d83163b38b1ecR100 is in this category  Any stuff belonging to this category can goes first
* settings related or dependent stuff
* everything else

Generally speaking this method seems to me quite long that does not help about understanding what is going one.

Comment (asmorodskyi) [reply openqabot/types/incidents.py:107]: any chance to do it in separate PR ?  we introducing quite big functional change here . IMHO it would be nice not to mix this with ANY other changes , so in case something will go wrong we can more easily detect the problem . If you will fully agree that this code may and should be improved . I just don't like to mix this **refactoring** together with introduction of `flavor_settings`

**Review by asmorodskyi — APPROVED** [review]
LGTM

File: openqabot/types/incidents.py
Comment (mpagot) [reply openqabot/types/incidents.py:103]: @asmorodskyi can we please further elaborate idea in
```
if "VERSION" in data["flavor_settings"] or "DISTRI" in data["flavor_settings"]:
                    raise ValueError("Attempt to redifine VERSION/DISTRI on flavor level")
```

There's no other `raise` in that function. Should we do it and stop everything else, every other `flavors`? Can we just `continue` and skip scheduling  `POST` for the "malicious flavor"? Can we simple remove any `DISTRI` and `VERSION` from `flavor_settings` before to merge it to the `full_post["openqa"]` ?

File: openqabot/types/incidents.py
Comment (asmorodskyi) [inline openqabot/types/incidents.py:109]: I would add some message here

**Review by alvarocarvajald — APPROVED** [review]
LGTM

**Review by asmorodskyi — APPROVED** [review]
LGTM

**Comment by baierjan** [comment]
Just out of curiosity, is the usage somewhat similar to the `params_expand` parameter?

**Comment by mpagot** [comment]
> `params_expand`

:face_palm: seems mostly or exactly the same

`param_expand` is there from the first commit in github, it has been cited in an example of the configure.md but nothing more. There's no unit testing for that. Apart from that, looking at the code, seems also to me that is capable to do exactly the same of this PR is proposing for `flavor_settings`. One missing feature is possibly the one to filter `DISTRI` and `VERSION`

**Comment by mpagot** [comment]
can we consider to merge this one and try to unify the two in a next PR?

**Review by baierjan — APPROVED** [review]
It's OK I guess.

File: openqabot/types/incidents.py
Comment (okurz) [inline openqabot/types/incidents.py:239]: the indendation level seems to become excessive here. Could you please try to extract this and the neighboring code parts into a function so that the level of indendation can be reduced here?

**Review by okurz — CHANGES_REQUESTED** [review]

File: openqabot/types/incidents.py
Comment (mpagot) [reply openqabot/types/incidents.py:239]: `black` does not blame me about it

Comment (alvarocarvajald) [reply openqabot/types/incidents.py:239]: @okurz can you unblock this merge? There already 2 approvals and @mpagot will address your suggestion in an upcoming PR (we're not even close to being done with this task)

Comment (okurz) [reply openqabot/types/incidents.py:239]: > But it means a big change, big just from the point of view of how many lines of code are moved around.
> I propose to do it in a next PR.
> How it sound?

Sorry, but this should be done as part of this PR. Make it a separate commit, sure, but still part of this PR. Unfortunately I know how often it happens that after a change someone has no time to work on promised changes anymore and then the codebase deteriorates and invites for more mistakes by others. Maybe it wouldn't happen here but that's not an unusual experience. So please look into this before merging.

Comment (okurz) [reply openqabot/types/incidents.py]: Nevermind. As baierjan stated "It's OK I guess." :)

**Review by okurz — APPROVED** [review]

**Pull request merged**
